### PR TITLE
upgrade to Python 3.12

### PR DESCRIPTION
**Pull request by joeyorlando**
# What this PR does

Upgrade to Python 3.12 + fix several invalid test assertions that lead to test failures in the latest version of `pytest`:
```
AttributeError: 'called_once_with' is not a valid assertion. Use a spec for the mock if 'called_once_with' is meant to be an attribute.. Did you mean: 'assert_called_once_with'?
```

## Checklist

- [ ] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

### CODE REVIEW

File: engine/apps/integrations/tests/test_views.py [hunk 1]
```diff
@@ -4,17 +4,20 @@
...
 from apps.alerts.models import AlertReceiveChannel
 from apps.integrations.mixins import AlertChannelDefiningMixin
 
 
-class DatabaseBlocker(_DatabaseBlocker):
+class DatabaseBlocker(DjangoDbBlocker):
     """Customize pytest_django db blocker to raise OperationalError exception."""
 
+    def __init__(self, *args) -> None:
+        super().__init__(_ispytest=True)
```
Comment (joeyorlando) [inline engine/apps/integrations/tests/test_views.py:19]: without this we get:
```
def __init__(self, *, _ispytest: bool = False) -> None:
        if not _ispytest:  # pragma: no cover
>           raise TypeError(
                "The DjangoDbBlocker constructor is private. "
                "use the django_db_blocker fixture instead."
            )
E           TypeError: The DjangoDbBlocker constructor is private. use the django_db_blocker fixture instead.
```

File: engine/apps/schedules/models/on_call_schedule.py [hunk 2]
```diff
@@ -500,7 +500,7 @@ def refresh_ical_final_schedule(self):
                     # check if event was ended or cancelled, update ical
                     dtend = component.get(ICAL_DATETIME_END)
                     dtend_datetime = dtend.dt if dtend else None
-                    if dtend_datetime and type(dtend_datetime) == datetime.date:
+                    if dtend_datetime and type(dtend_datetime) is datetime.date:
```
Comment (joeyorlando) [inline engine/apps/schedules/models/on_call_schedule.py:503]: the bump in `black` from `23.7.0` to `23.11.0` was complaining that these type assertions should instead use `is` / `is not`

File: engine/requirements.txt [hunk 3]
```diff
@@ -7,7 +7,7 @@ phonenumbers==8.10.0
 celery[amqp,redis]==5.3.1
 redis==5.0.1
 humanize==0.5.1
-uwsgi==2.0.21
+uwsgi==2.0.23
```
Comment (vadimkerr) [inline engine/requirements.txt:10]: might need to check https://github.com/grafana/oncall-private/issues/2125

Comment (joeyorlando) [reply engine/requirements.txt]: this upgrade was performed in [v1.3.67](https://github.com/grafana/oncall/releases/tag/v1.3.67) 👍

**Review by vadimkerr — APPROVED** [review]

**Pull request merged**